src/main/java/net/floodlightcontroller/statistics/StatisticsCollector.java

```diff
@@ -162,6 +171,107 @@ protected long getSpeed(NodePortTuple npt) {
 		}
 
 	}
+	
+	
+	/*
+	 * 	Added by Charlie Hardwick-Kelly Carleton University 
+	 * 	The purpose of this class is to track per flow metrics from within the network  
+	 * 	one key advantage of the flow metrics is that they contain a duration time stamp 
+	 * 	this eliminates the error involved from the control plane and allows us to compute the bandwidth of 
+	 * 	a particular flow with greater accuracy
+	 * 
+	 *  	Steps 
+	 * 		1. issue a meter stats request to all switches in parallel 
+	 * 		2. Analyze the variables 
+	 * 		3. if valid calculate relevant bandwidth 
+	 * 		4. store information in hash 
+	 * 		5. send information to the gui.
+	 * 
+	 */
+	protected class FlowStatsCollector implements Runnable {
+
+		@Override
+		public void run() {
+			Map<DatapathId, List<OFStatsReply>> flowReplies = getSwitchStatistics(switchService.getAllSwitchDpids(), OFStatsType.FLOW);
+			if(debug) {
+				System.out.print("\033[H\033[2J");// add this to clear the command line every round only works on real terminal
+			}
+			for (Entry<DatapathId, List<OFStatsReply>> e : flowReplies.entrySet()) {
+				for (OFStatsReply r : e.getValue()) {
+
+					OFFlowStatsReply fsr = (OFFlowStatsReply) r; 
+					for ( OFFlowStatsEntry fse : fsr.getEntries()) {
+						//extract the necessary information from the existing stats and 
+						DatapathId sw = e.getKey();
+						Match match = fse.getMatch(); 						
+						OFPort in_port = match.get(MatchField.IN_PORT);
+						U64 bytesCount = fse.getByteCount(); //take in the number of bytes  
+						long dur = fse.getDurationSec(); //take in the duration 
+						// now I need to set up the comparisons 
+						NodeFlowTuple nft = new NodeFlowTuple(sw,match) ;
+						if(flowStats.containsKey(nft)){	
+							//Retrieve the previous stat response information from the hash
+							FlowBandwidth stat = flowStats.get(nft);
+							
+							//Calculate the bytes difference between current and previous collections
+							// convert from bytes to bits 
+							long byteDiff = (bytesCount.getValue() - stat.getBytes().getValue())*BITS_PER_BYTE;
+							
+							//Calculate the difference in time between the current and previous stat response
+							long timediff = dur - stat.getDuration();
+							
+							long speed;
+							// Avoid divide by zero error
+							if(byteDiff!=0){							
+								speed= byteDiff/timediff;
+							}else {
+								speed= 0;
+							}
+							try {
+								if(!(speed < 0)) {		
+									// if we have a valid flow we want to update the hash and then send the information to the GUI
+									
+									
+									FlowBandwidth test = FlowBandwidth.of(sw, match, dur, bytesCount, speed,in_port);
+									flowStats.put(nft,test);		
+									
+									//The SOCKET send to the gui should be implemented here 
```

Comment:
Is this how we were planning on sending flow data to the GUI?

Reply:
I wasn't sure, but if we want to send the data as soon as its created thats where we'll need to do it